buildSrc/src/main/kotlin/BuildInfo.kt

```diff
@@ -19,7 +19,34 @@ import java.io.File
 import org.gradle.api.Project
 import org.gradle.api.artifacts.VersionCatalog
 import org.gradle.api.artifacts.VersionCatalogsExtension
+import org.gradle.api.provider.Provider
+import org.gradle.jvm.toolchain.*
 import org.gradle.kotlin.dsl.getByType
+import org.gradle.kotlin.dsl.support.serviceOf
+
+/**
+ * JVM bytecode target; this is pinned at a reasonable version, because downstream JVM projects
+ * which consume Pkl will need a minimum Bytecode level at or above this one.
+ *
+ * Kotlin and Java need matching bytecode targets, so this is expressed as a build setting and
+ * constant default. To override, pass `-PpklJvmTarget=X` to the Gradle command line, where X is a
+ * major Java version.
+ */
+private const val PKL_JVM_TARGET_DEFAULT_MAXIMUM = 17
+
+/**
+ * The Pkl build requires JDK 21+ to build, because JDK 17 is no longer within the default set of
+ * supported JDKs for GraalVM. This is a build-time requirement, not a runtime requirement.
+ *
+ * See `settings.gradle.kts`, where this is enforced.
+ */
+private const val PKL_JDK_VERSION_MIN = 21
+
+/**
+ * The Pkl build must use exactly JDK 21 to run tests, because certain methods in use by Truffle are
+ * not present in JDK 22+; thus, this is pinned to [PKL_JDK_VERSION_MIN] (JDK 21).
+ */
+private const val PKL_TEST_JDK_TARGET = PKL_JDK_VERSION_MIN
```

Comment:
this now documents the JVM bytecode level, JDK build level, and JDK/JVM test level, and the rationale about why they are set the way they are.